### ISSUE TITLE
🚀 Feature: Running parallel tests in isolation

### ISSUE THREAD

**Issue by rochdev** [mochajs/mocha#4947]
**Is your feature request related to a problem or a nice-to-have?? Please describe.**

Right now `mocha --parallel` works by using a specific number of workers and running tests in parallel across these workers. This means a single worker ends up running many test suites in the same process. The main reason we wanted to parallelize tests is not for speed but to isolate tests, which is not the case even in parallel.

**Describe the solution you'd like**

It would be great to have a mode where the tests could be completely isolated so that a new worker is started for every file, thus avoiding side-effects from previous files.

**Describe alternatives you've considered**

Usually, the alternative is to write code and tests in a way that avoids side-effects so that every test can be run in the same process without issues. However, for our case, we patch and instrument many things that are irreversible, like enabling `async_hooks`, patching the module system, patching dependencies as they are imported, adding ESM loaders, etc. This makes it basically impossible to avoid side-effects, so our only option is to ensure test isolation.

**Additional context**

I wasn't able to find a way to do this with the latest version of Mocha, but please let me know if there is already a way to do this.

**Comment by rochdev** [comment]
It looks like setting the maximum number of jobs higher than the number of files result in the isolation we're looking for (or at least it looks like it). However, this means spawning dozens if not hundreds of workers simultaneously.

**Comment by JoshuaKGoldberg** [comment]
Yeah I can see why you'd want this. Newer test runners such as Jest and Vitest do this automatically, and it's a feature gap of Mocha to not be able to at least opt into this. I think this is a useful enough feature to warrant tackling eventually.

Note that per #5027 we're trying to take it slow on any big changes. Even if this ends up being a small opt-in feature, it still touches enough very tricky code that we likely won't get to it for a while. Even reviewing a PR for it would likely be blocked on us getting through many existing issues & PRs.

cc @mochajs/maintenance-crew - thoughts?

**Comment by voxpelli** [comment]
Not sure I think this is in scope of what Mocha should do.

As you say @JoshuaKGoldberg, eg. Jest and Vitest do this automatically, so if one has a critical need for this kind of functionality one can pick one of them.

Another alternative could be to eg. invoke the `mocha` cli multiple times in parallell and to merge the results of all such invocations, then it for sure would be isolated.

I'm 👎 for now.

If there needs to be a clarification of the limits of isolation in parallelization docs then we can look at that.

(Side note, this reminds me of #1948 about the `--check-leaks`)

**Comment by JoshuaKGoldberg** [comment]
👍 the lack of significant engagement on this issue or consensus from maintainers makes me agree, it's not a priority for the project. 5 upvotes across ~2.5 years of this issue existing = 1 person asking every ~5 months for it. Not a lot of signaled interest. +1 to pointing folks to other frameworks like Jest and Vitest that are architected for the isolation use case. 

Filed #5356 separately for docs.
